File: frontend/src/lib/rut.ts
/**
 * Utilidades para RUT chileno (módulo 11).
 */
export function validarRut(rut: string): boolean {
  const clean = normalizarRut(rut);
  if (!clean.includes("-")) return false;
  const [body, dv] = clean.split("-");
  if (!/^\d+$/.test(body) || body.length < 7) return false;
  let s = 0, m = 2;
  for (const c of body.split("").reverse()) {
    s += parseInt(c) * m;
    m = m < 7 ? m + 1 : 2;
  }
  const expected = 11 - (s % 11);
  const expectedDv = expected === 11 ? "0" : expected === 10 ? "K" : String(expected);
  return dv === expectedDv;
}

export function normalizarRut(rut: string): string {
  return rut.replace(/\./g, "").replace(/\s/g, "").toUpperCase().trim();
}

export function formatearRut(rut: string): string {
  const clean = normalizarRut(rut);
  if (!clean.includes("-")) return clean;
  const [body, dv] = clean.split("-");
  let formatted = "";
  for (let i = 0; i < body.length; i++) {
    if (i > 0 && (body.length - i) % 3 === 0) formatted += ".";
    formatted += body[i];
  }
  return `${formatted}-${dv}`;
}

File: anamnesia/app/utils/rut.py
"""Utilidades para validación y formateo de RUT chileno (módulo 11)."""

def validar_rut(rut: str) -> bool:
    clean = normalizar_rut(rut)
    if "-" not in clean:
        return False
    body, dv = clean.rsplit("-", 1)
    if not body.isdigit() or len(body) < 7:
        return False
    s, m = 0, 2
    for c in reversed(body):
        s += int(c) * m
        m = m + 1 if m < 7 else 2
    expected = 11 - (s % 11)
    if expected == 11:
        expected_dv = "0"
    elif expected == 10:
        expected_dv = "K"
    else:
        expected_dv = str(expected)
    return dv == expected_dv

def normalizar_rut(rut: str) -> str:
    return rut.replace(".", "").replace(" ", "").upper().strip()

def formatear_rut(rut: str) -> str:
    clean = normalizar_rut(rut)
    if "-" not in clean:
        return clean
    body, dv = clean.rsplit("-", 1)
    formatted = ""
    for i, c in enumerate(reversed(body)):
        if i > 0 and i % 3 == 0:
            formatted = "." + formatted
        formatted = c + formatted
    return f"{formatted}-{dv}"
